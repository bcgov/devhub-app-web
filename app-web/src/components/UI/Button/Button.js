import React from 'react';
import PropTypes from 'prop-types';
import classes from './Button.module.css';
import { buttonTypes } from '../../../constants/ui';

const Button = ({ type, children, clicked, ...rest }) => {
  let buttonClasses = [classes.Button];
  const typeClass = classes[type];
  buttonClasses = buttonClasses.concat([typeClass]);
  return (
    <button className={buttonClasses.join(' ')} onClick={clicked} {...rest}>
      {children}
    </button>
  );
};

Button.propTypes = {
  children: PropTypes.string,
  clicked: PropTypes.func.isRequired,
  type: PropTypes.oneOf(buttonTypes),
};

Button.defaultProps = {
  children: '',
};

export default Button;
