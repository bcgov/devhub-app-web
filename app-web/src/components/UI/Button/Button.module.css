.Button {
    /* taken from gov bootstrap template https://github.com/bcgov/Gov-2.0-Bootstrap-Skeleton */
    display: inline-block;
    font-weight: 400;
    text-align: center;
    white-space: nowrap;
    vertical-align: middle;
    -webkit-user-select: none;
    -moz-user-select: none;
    -ms-user-select: none;
    user-select: none;
    border: 1px solid transparent;
    padding: .375rem .75rem;
    font-size: 1em;
    line-height: 1.5;
    border-radius: .25rem;
    cursor: pointer;
}

.Button:focus {
    outline: none;
}
/* these classes aren't capatalized for ease of using these in dynamic logic */
.Button.primary {
    background-color: #002663;
    color: #fff;
}

.Button.secondary {
    background-color: #E3A82B;
    color: #fff;
}

.Button.link {
    background-color: transparent;
    outline: none;
    color: inherit;
}
