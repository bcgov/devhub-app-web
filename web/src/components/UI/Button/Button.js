import React from 'react';
import PropTypes from 'prop-types';
import styled from '@emotion/styled';
import rocketchatSVG from '../../../assets/images/rocketchat_logo.svg';
import githubPNG from '../../../assets/images/github_logo.png';
import DocumizePNG from '../../../assets/images/documize_logo.png';
import { SEARCH_SOURCES } from '../../../constants/search';

const StyledButton = styled.button`
  color: #fff;
  user-select: none;
  border: 1px solid transparent;
  padding: 0.375rem 0.75rem;
  font-size: 1em;
  line-height: 1.5;
  border-radius: 0.25rem;
  cursor: pointer;
  font-weight: 400;
  text-align: center;
  white-space: nowrap;
  vertical-align: middle;
  display: inline-block;
  background-color: ${({ color, theme }) => {
    if (theme.colors[color]) {
      return theme.colors[color];
    } else {
      return 'initial';
    }
  }};
`;

const variants = {
  primary: 'primary',
  secondary: 'secondary',
  link: 'link',
};

export const TEST_IDS = {
  button: 'button',
};

export const Button = ({ variant, children, clicked, ...rest }) => {
  const colors = {
    [variants.primary]: 'blue',
    [variants.secondary]: 'yellow',
  };
  if (variant === variants.link) {
    return (
      <StyledButton
        onClick={clicked}
        style={{
          backgroundColor: 'transparent',
          outline: 'none',
          color: 'inherit',
        }}
        data-testid={TEST_IDS.button}
        {...rest}
      >
        {children}
      </StyledButton>
    );
  } else {
    return (
      <StyledButton
        onClick={clicked}
        color={colors[variant]}
        data-testid={TEST_IDS.button}
        {...rest}
      >
        {children}
      </StyledButton>
    );
  }
};

Button.propTypes = {
  children: PropTypes.node,
  clicked: PropTypes.func.isRequired,
  variant: PropTypes.oneOf(Object.keys(variants)),
};

Button.defaultProps = {
  children: '',
};

export const IconButton = styled.button`
  background-color: transparent;
  border: none;
  cursor: pointer;
  width: 35px;
  padding: 0;
  :focus,
  :active {
    outline: none;
  }
  :disabled {
    opacity: 0.65;
  }
`;

const SEARCH_SOURCES_ICON = {
  [SEARCH_SOURCES.rocketchat]: rocketchatSVG,
  [SEARCH_SOURCES.github]: githubPNG,
  [SEARCH_SOURCES.documize]: DocumizePNG,
};

export const SearchSourcesButton = props => (
  <IconButton {...props}>
    <img
      src={SEARCH_SOURCES_ICON[props.searchType]}
      style={{ margin: 0 }}
      alt={props.searchType + ' logo'}
    />
  </IconButton>
);

SearchSourcesButton.propTypes = {
  searchType: PropTypes.string.isRequired,
  style: PropTypes.object,
  title: PropTypes.string.isRequired,
};

export default Button;
