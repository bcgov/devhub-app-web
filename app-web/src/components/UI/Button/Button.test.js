import React from 'react';
import { shallow } from 'enzyme';
import Button from './Button';

describe('Button Component', () => {
  it('matches snapshot', () => {
    const wrapper = shallow(<Button clicked={jest.fn()} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('it calls event handling function when clicked', () => {
    const clickFn = jest.fn();
    const wrapper = shallow(<Button clicked={clickFn} />);
    wrapper.simulate('click');
    expect(clickFn).toHaveBeenCalled();
  });
});
