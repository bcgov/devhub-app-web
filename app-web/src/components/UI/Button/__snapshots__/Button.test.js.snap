// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Component matches snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Button
    clicked={[Function]}
  >
    
  </Button>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": "",
      "className": "Button undefined",
      "onClick": [Function],
    },
    "ref": null,
    "rendered": "",
    "type": "button",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "",
        "className": "Button undefined",
        "onClick": [Function],
      },
      "ref": null,
      "rendered": "",
      "type": "button",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
